src/sentry/notifications/notificationcontroller.py

```diff
@@ -0,0 +1,460 @@
+            return (NotificationScopeEnum.USER, [recipient])
+        if recipient_is_team(recipient):
+            return (NotificationScopeEnum.TEAM, [recipient])
+
+        raise ValueError(f"Recipient is of unsupported type: {recipient}")
+
+    def _filter_options(
+        self,
+        settings: Iterable[NotificationSettingOption],
+        **kwargs,
+    ) -> list[NotificationSettingOption]:
+        return [
+            setting
+            for setting in settings
+            if all(getattr(setting, arg) == kwargs[arg] for arg in kwargs)
+        ]
+
+    def _filter_providers(
+        self,
+        settings: Iterable[NotificationSettingProvider],
+        **kwargs,
+    ) -> list[NotificationSettingProvider]:
+        return [
+            setting
+            for setting in settings
+            if all(getattr(setting, arg) == kwargs[arg] for arg in kwargs)
+        ]
+
+    def _get_layered_setting_options(
+        self,
+        **kwargs,
+    ) -> MutableMapping[
+        Recipient,
+        MutableMapping[
+            Scope, MutableMapping[NotificationSettingEnum, NotificationSettingsOptionEnum]
+        ],
+    ]:
+        """
+        Returns a mapping of the most specific notification setting options for the given recipients and scopes.
+        Note that this includes default settings for any notification types that are not set.
+
+        Args:
+            setting_type: If specified, only return settings of this type.
+        """
+        settings = self.get_all_setting_options()
+        project_settings = self._filter_options(
+            settings=settings, scope_type=NotificationScopeEnum.PROJECT.value, **kwargs
+        )
+        org_settings = self._filter_options(
+            settings=settings, scope_type=NotificationScopeEnum.ORGANIZATION.value, **kwargs
+        )
+        user_settings = self._filter_options(
+            settings=settings, scope_type=NotificationScopeEnum.USER.value, **kwargs
+        )
+        scoped_settings = [project_settings, org_settings, user_settings]
+
+        defaults = get_type_defaults()
+        layered_setting_options: MutableMapping[
+            Recipient,
+            MutableMapping[
+                Scope, MutableMapping[NotificationSettingEnum, NotificationSettingsOptionEnum]
+            ],
+        ] = defaultdict(lambda: defaultdict(dict))
+        for recipient in self.recipients:
+            (scope_type, scope_items) = self._get_scope_for_layering(recipient)
+            for item in scope_items:
+                scope: Scope = (scope_type, item)
+                for type in NotificationSettingEnum:
+                    most_specific_setting = None
+                    for setting in scoped_settings:
+                        user_settings = self._filter_options(
+                            settings=setting, user_id=recipient.id, type=type.value
+                        )
+                        team_settings = self._filter_options(
```

Comment:
@snigdhas I think this will result in bugs unless you check the type of the recipient because it's possible the ids of a team will match a user and vice-versa.